src/url-replacements.js

```diff
@@ -111,15 +119,42 @@ class UrlReplacements {
    */
   expand(url, opt_data) {
     const expr = this.getExpr_();
-    return url.replace(expr, (match, name) => {
-      let val = this.replacements_[name](opt_data);
+    let replacementPromise;
+    const encodeValue = val => {
       // Value 0 is specialcased because the numeric 0 is a valid substitution
       // value.
       if (!val && val !== 0) {
         val = '';
       }
       return encodeURIComponent(val);
+    };
+    url = url.replace(expr, (match, name, arg) => {
+      const val = this.replacements_[name](opt_data, arg);
+      // In case the produced value is a promise, we don't actually
+      // replace anything here, but do it again when the promise resolves.
+      if (val instanceof Promise) {
+        const p = val.then(v => {
+          url = url.replace(match, encodeValue(v));
```

Comment:
This has a slight negative that two RANDOM would produce the same value when replaced. But this is somewhat of a contrived example. Alternatively, we could insert some autogenerated match string (REPLACE_ME_1,2,3) and promises when return would put values into this map and after all promises returned, it would do one set of replacements from this map. But, if it doesn't seem like an issue, we can just keep as is.